chain/sub/bcast/consistent.go

```diff
@@ -0,0 +1,180 @@
+package bcast
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/multiformats/go-multihash"
+
+	"github.com/filecoin-project/go-state-types/abi"
+
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+var log = logging.Logger("sub-cb")
+
+const (
+	// GcSanityCheck determines the number of epochs that in the past
+	// that will be garbage collected from the current epoch.
+	GcSanityCheck = 5
+	// GcLookback determines the number of epochs kept in the consistent
+	// broadcast cache.
+	GcLookback = 2
+)
+
+type blksInfo struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	blks   []cid.Cid
```

Comment:
I'm unsure about the thread-safety of this `blks` slice. We seem to be reading & modifying it somewhat unsafely, and while it's probably the case that the system won't ever race here, I'm...not sure.

Can we reason through this a bit?

Reply:
Actually, looking at the code closer, we don't actually ever modify it. So we're probably okay here.

Reply:
@adlrocha I don't think that's sufficient. I think it needs to be held until no more writing happens, which is the end of the method (as currently written). Consider the following situation:

- 2 blocks A and B come in with the same VRFProof. 
- A gets the lock first, creates an empty BCastDict (line 119), doesn't find a match for the common `key` on line 125, drops the lock
- B gets the lock, loads the empty BCastDict created by A, also doesn't find a match for the common `key` on line 125, drops the lock
- A `store`s `blkACid` as the only entry corresponding to `key` on line 141, returns
- B _overwrites_ that, storing `blkBCid` as the _only_ entry corresponding to `key` on line 141, returns
- Both A and B succeed their respective calls to `WaitForDelivery`